fulltext/extract.py

```diff
@@ -0,0 +1,66 @@
+"""Provides asynchronous task for fulltext extraction."""
+
+from celery import shared_task
+import os
+from datetime import datetime
+from fulltext import logging
+from fulltext.services.fulltext import extractor
+from fulltext.services.store import store
+from fulltext.services.metrics import metrics
+from fulltext.services.retrieve import retrievePDF
+from celery.result import AsyncResult
+from celery import current_app
+from celery.signals import after_task_publish
+logger = logging.getLogger(__name__)
+
+
+@shared_task
+def extract_fulltext(document_id: str, pdf_url: str) -> None:
+    """Perform fulltext extraction for a single arXiv document."""
+    logger.info('Retrieving PDF for %s' % document_id)
+    start_time = datetime.now()
+    try:
+        # Retrieve PDF from arXiv central document store.
+        pdf_path = retrievePDF.session.retrieve(pdf_url, document_id)
+        if pdf_path is None:
+            metrics.session.report('PDFIsAvailable', 0.)
+            msg = '%s: no PDF available' % document_id
+            logger.info(msg)
+            raise RuntimeError(msg)
+        metrics.session.report('PDFIsAvailable', 1.)
+        logger.info('%s: retrieved PDF' % document_id)
+
+        logger.info('Attempting text extraction for %s' % document_id)
+        txt_path = extractor.session.extract_fulltext(pdf_path)
+        logger.info('Text extraction for %s succeeded with %s' %
+                    (document_id, txt_path))
+
+        os.remove(pdf_path)    # Cleanup.
+        with open(txt_path, encoding='utf-8') as f:
+            content = f.read()
```

Comment:
Do we know that reading the whole file will fit safely in memory?

Reply:
That's a great question; good to think through. 

Academic papers tend to be in the several- to dozens-of-pages range, but let's suppose that someone uploads a book manuscript to arXiv that is 500 pages long. A single-spaced page at 12pt font might have 500 words; assuming an average word length of 5 characters, that's 2,500 characters per page. 2,500 char/page * 500 pages = 1.25 million characters. In an extreme case, we might have an average of 2 bytes per character -- suppose someone making heavy use of accents or other non-ASCII symbols. So 2.5 megabytes. 

The bigger issue will be the PDF. Assume that we're generating PDFs that are 100kb/page (high, but suppose that there are figures) -- that's a 50MB PDF in our extreme example. Still a pretty reasonable number. 

As we get a bit further along with Kubernetes, we'll want to make these requirements more explicit.